tests/test_edge_cases.py

```diff
@@ -1506,3 +1506,42 @@ class Model(BaseModel):
         'nullable1': 1,
         'nullable2': 'two',
     }
+
+
+def test_custom_generic():
+    T1 = TypeVar('T1')
+    T2 = TypeVar('T2')
+
+    class MyGen(Generic[T1, T2]):
+        def __init__(self, t1: T1, t2: T2):
+            self.t1 = t1
+            self.t2 = t2
+
+        @classmethod
+        def __get_validators__(cls):
+            yield cls.validate
+
+        @classmethod
+        def validate(cls, v):
+            if not isinstance(v, cls):
+                raise ValueError('Invalid value')
```

Comment:
Do `isinstance(v.t1, cls.__args__[0])` and `isinstance(v.t2, cls.__args__[1])` work here? I think it might be good to have an example that actually checks the generic types in the validators, for reference by others.

That could just go in the docs, but putting it in the tests too will help make sure it actually works 😅.

(In general you'd want to handle some edge cases better maybe, but for a test/demonstration I think this would be fine.)

Reply:
Good point, I'm not sure it actually works, because the `Field.type_` that is originally the generic is a type alias and the `origin` is the one that has the `__get_validators__`, but I'll try to see if it works.

Reply:
Hmm, I think it would be unfortunate if the implementation *prevented* you from actually checking the specific parameter types, but maybe that can be fixed later if necessary.

Reply:
Yeah, just checked.

Sadly, for `validate` and `__get_validators__`, `cls` is a `<class '__main__.MyGen'>`, not a `__main__.MyGen[str, bool]`.

The way we could implement it would be to send the sub types to the validators provided by `__get_validators__` as extra arguments, during validation.

But that would break the current API. I guess it would require another method.

Or, another option is to inspect each validator function returned to see if it requests the types. But that adds extra complexity and I'm not sure how much should we add for this version at least. That option would probably have some performance penalty too, so it might be better just as another method.

Reply:
It should be possible using the `field` argument which can optionally be passed to a validator?

Reply:
Yep, makes sense.

I think you could do something tricky with signature inspection of `__get_validators__` to optionally allow an extra keyword argument in the `__get_validators__` function which would receive the field type. (This is how `@validator` works now with the functions it decorates now, I believe.)

Not sure if it's worth the complexity, but on principle I like it.

Reply:
Yeah, for performance reasons I think probably it should be the `__get_validators__` function that receives the field type, and then that function yields "configured" validators.

Reply:
`field` is already available as an argument to a valdator, it will only be provided if the validator takes that argument.

Are you sure it's not possible now?

Reply:
I hadn't seen these comments before writing my last comment down below. :sweat_smile: 

I'm realizing that `__get_validators__` is called on creation, so having some small extra logic for it shouldn't be too bad, as it's called only once on creation, during `populate_validators()`.

Given that, I think I can make it accept an optional `FieldModel`, very much like in FastAPI (almost everywhere).

The API would look like this:

if `__get_validators__` has a signature like:

```Python
@classmethod
def __get_validators__(cls, field: FieldModel):
    ...
```

Then it would receive the `FieldModel` in the `field` argument.

Otherwise, it would work as normal (no arguments).

Does that sound okay?

Reply:
Yeah, given that this is already implemented for validators that seems like the way to go! Awesome!

Reply:
```suggestion
                raise TypeError('Invalid value')
```